pkg/providers/cloudstack/validator.go

```diff
@@ -48,83 +55,122 @@ type ProviderCmkClient interface {
 	return nil
 }
 
-func (v *Validator) validateDomainAndAccount(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
-	if (datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account == "") ||
-		(datacenterConfig.Spec.Domain == "" && datacenterConfig.Spec.Account != "") {
-		return fmt.Errorf("both domain and account must be specified or none of them must be specified")
+func (v *Validator) generateLocalAvailabilityZones(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
+	if datacenterConfig == nil {
+		return nil
 	}
 
-	if datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account != "" {
-		domain, errDomain := v.cmk.ValidateDomainPresent(ctx, datacenterConfig.Spec.Domain)
-		if errDomain != nil {
-			return fmt.Errorf("checking domain: %v", errDomain)
+	if len(datacenterConfig.Spec.Domain) > 0 {
+		cmk := v.cmks[decoder.CloudStackGlobalAZ]
+		domain, err := cmk.ValidateDomainPresent(ctx, datacenterConfig.Spec.Domain)
+		if err != nil {
+			return err
 		}
-
-		errAccount := v.cmk.ValidateAccountPresent(ctx, datacenterConfig.Spec.Account, domain.Id)
-		if errAccount != nil {
-			return fmt.Errorf("checking account: %v", errAccount)
+		if err := cmk.ValidateAccountPresent(ctx, datacenterConfig.Spec.Account, domain.Id); err != nil {
+			return err
+		}
+		for _, zone := range datacenterConfig.Spec.Zones {
+			availabilityZone := localAvailabilityZone{
+				CloudStackAvailabilityZone: &anywherev1.CloudStackAvailabilityZone{
+					Name:                  decoder.CloudStackGlobalAZ,
+					Domain:                datacenterConfig.Spec.Domain,
+					Account:               datacenterConfig.Spec.Account,
+					ManagementApiEndpoint: datacenterConfig.Spec.ManagementApiEndpoint,
+					Zone:                  zone,
+				},
+				DomainId: domain.Id,
+			}
+			v.availabilityZones = append(v.availabilityZones, availabilityZone)
+		}
+	}
+	for _, az := range datacenterConfig.Spec.AvailabilityZones {
+		cmk := v.cmks[az.Name]
+		domain, err := cmk.ValidateDomainPresent(ctx, az.Domain)
+		if err != nil {
+			return err
+		}
+		if err := cmk.ValidateAccountPresent(ctx, az.Account, domain.Id); err != nil {
+			return err
 		}
+		availabilityZone := localAvailabilityZone{
+			CloudStackAvailabilityZone: &az,
+			DomainId:                   domain.Id,
+		}
+		v.availabilityZones = append(v.availabilityZones, availabilityZone)
+	}
 
-		domainId = domain.Id
+	if len(v.availabilityZones) <= 0 {
+		return fmt.Errorf("CloudStackDatacenterConfig domain or availabilityZones is not set or is empty")
 	}
 	return nil
 }
 
 // TODO: dry out machine configs validations
 func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, cloudStackClusterSpec *Spec) error {
-	var etcdMachineConfig *anywherev1.CloudStackMachineConfig
+	if err := v.ValidateCloudStackDatacenterConfig(ctx, cloudStackClusterSpec.CloudStackDatacenter); err != nil {
+		return err
+	}
```

Comment:
why do we want to do this? It'll make testing harder, and we're already calling this method from cloudstack.go. This way it's harder to tell which method is misbehaving if there's a dependency between them.